File: packages/factory/src/core/context.rs
use swc_core::common::sync::Lrc;
use swc_core::common::{comments::Comments, SourceFile};

use crate::core::messages::CompositeMessage;

/// Context for the visitor, used to share and accumulate data during AST traversal or analysis.
///
/// This includes immutable data like the source file and comment references,
/// as well as mutable data like collected messages and a counter-based identifier store.
pub struct Context {
  pub source_file: Option<Lrc<SourceFile>>,
  pub single_comments: Option<Lrc<dyn Comments>>,
  pub found_messages: Vec<CompositeMessage>,
  pub identifier_store: IdentifierStore,
}
impl Context {
  pub fn new(
    source_file: Option<Lrc<SourceFile>>,
    single_comments: Option<Lrc<dyn Comments>>,
  ) -> Context {
    Context {
      source_file,
      single_comments,
      found_messages: vec![],
      identifier_store: IdentifierStore::default(),
    }
  }
}

/// A store for generating simple numeric-based unique string identifiers.
///
/// Useful for tracking unnamed entities or generating consistent IDs across visits.
#[derive(Default)]
pub struct IdentifierStore {
  current: usize,
}
impl IdentifierStore {
  pub fn next(&mut self) -> String {
    let id = self.current;
    self.current += 1;
    id.to_string()
  }
  pub fn back(&mut self) {
    self.current = self.current.saturating_sub(1);
  }
  pub fn reset(&mut self) {
    self.current = 0;
  }
}

File: packages/plugin-babel/src/core/context.ts
import type { Comment } from '@babel/types';
import type { CompositeMessage } from './messages/types.js';

export class Context {
  foundMessages: CompositeMessage[] = [];
  identifierStore = new IdentifierStore();
  constructor(public readonly comments: Comment[] | null | undefined) {}
}

export class IdentifierStore {
  #current = 0;

  next() {
    const id = this.#current;
    this.#current += 1;
    return id.toString();
  }

  back() {
    this.#current -= 1;
  }

  reset() {
    this.#current = 0;
  }
}
